### PR TITLE
refactor(v2): computational backends

### PR DESCRIPTION
**Pull request by JohannesMessner**
Goals:

This introduces "computational backends" for organizing and grouping functionality that is framework specific, e.g. numpy or torch

### CODE REVIEW

File: docarray/computation/abstract_comp_backend.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    @abstractmethod
+    def stack(
+        tensors: Union[List['Tensor'], Tuple['Tensor']], dim: int = 0
+    ) -> 'Tensor':
+        """
+        Stack a list of tensors along a new axis.
+        """
+        ...
+
+    class Retrieval(ABC):
```
Comment (samsja) [inline docarray/computation/abstract_comp_backend.py:26]: why is this inside the other class ?

Comment (JohannesMessner) [reply docarray/computation/abstract_comp_backend.py:26]: just for further structuring/namespacing things. so all the metrics can go in one place, all the retrieval related stuff in another place, etc

File: docarray/typing/abstract_type.py [hunk 2]
```diff
@@ -13,6 +13,8 @@
 
 
 class AbstractType(BaseNode):
+    is_tensor = False  # change for tensor-like subclasses
```
Comment (samsja) [inline docarray/typing/abstract_type.py:16]: I am not sure to understand why we need this. Only object of type AbstractTensor will have this `is_tensor=True` right ? So why not just do
```python
if issubclass(type_, AbstractType):
    if issubclass(type_, AbstractTensor):
        smth_with_tensor()
    else:
        smth_without_tensor

```

Comment (JohannesMessner) [reply docarray/typing/abstract_type.py:16]: Yeah true, will revert

**Review by samsja — APPROVED** [review]

**Pull request merged**